tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,348 @@
+        && weightsShape.isCompatibleWith(valuesShape)) {
+      return weights;
+    }
+
+    Ops ctf =
+        tf.withSubScope("broadcastWeights")
+            .withControlDependencies(
+                Collections.singletonList(assertBroadcastable(tf, weights, tf.onesLike(values))));
+    return ctf.math.mul(weights, tf.onesLike(values));
+  }
+
+  // aliases for mean
+
+  /**
+   * Calculate the mean of the operand, along all axes and <code>keepDims</code> is <code>false
+   * </code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param <T> the type of the Operand.
+   * @return the mean of the operand
+   */
+  public static <T extends TNumber> Operand<T> mean(Ops tf, Operand<T> x) {
+    return mean(tf, x, null, false);
+  }
+
+  /**
+   * Calculate the mean of the operand, alongside the specified axis with <code>keepDims</code> is
+   * <code>false</code>
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axes Axes to compute the mean.
+   * @param <T> the type of the Operand.
+   * @param <U> the type of the axes.
+   * @return the mean of the operand, along the specified axes.
+   */
+  public static <T extends TNumber, U extends TIntegral> Operand<T> mean(
+      Ops tf, Operand<T> x, Operand<U> axes) {
+    return mean(tf, x, axes, false);
+  }
+
+  /**
+   * Calculates the mean of the operand, along all axes.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param keepDims Indicates whether to keep the dimensions or not. If <code>keepdims</code> is
+   *     <code>false</code>, the rank of the tensor is reduced by 1 for each entry in <code>axes
+   *     </code>. If <code>keepdims</code> is <code>true</code>, the reduced dimensions are retained
+   *     with length 1.
+   * @param <T> the type of the operand
+   * @return the mean of elements of <code>x</code>.
+   */
+  public static <T extends TNumber> Operand<T> mean(
+      Ops tf, Operand<T> x, boolean keepDims) {
+    return mean(tf, x, null, keepDims);
+  }
+
+
+
+  /**
+   * Calculates the mean of the operand, alongside the specified axes.
+   *
+   * @param tf the TensorFlow Ops
+   * @param x the Operand used to calculate the mean
+   * @param axes Axes to compute the mean.
+   * @param keepDims Indicates whether to keep the dimensions or not. If `keepdims` is `false`, the
+   *     * rank of the tensor is reduced by 1 for each entry in `axes`. If `keepdims` is `true`, the
+   *     * reduced dimensions are retained with length 1.
+   * @param <T> the data type of the Operand
+   * @param <U> the data type of the axes
+   * @return the mean of elements of <code>x</code>.
+   */
+
+  public static <T extends TNumber, U extends TIntegral> Operand<T> mean(
+      Ops tf, Operand<T> x, Operand<U> axes, boolean keepDims) {
+    if (axes == null) {
+      axes = (Operand<U>) allAxes(tf, x);
```

Comment:
I like the direction you've taken for the `mean` and `meanBoolean` methods! The cast on this line still has a problem: `allAxes` always returns an `Operand<TInt32>`, which may not be the type specified by the caller as `U`. I don't see `U` serving a purpose here. I think it would work to drop that type parameter and just declare `Operand<? extends TIntegral>` in the signature. Otherwise, perhaps split the `axes == null` process completely into a different `if` branch?

Reply:
We should try to use much wildcards as possible in the method signatures rather than generic parameters, please see this PR from @rnett that got rid of these superfluous parameters in the generated ops: #193

@JimClarke5 , do you think you can do a similar cleanup in your code?

Reply:
I haver started to look at it, and I think I can get most everything to one Generic Parameter, mainly for return type, but it will impact a lot of code including `losses`. I would not to do this now, but in a subsequent PR.

Reply:
Ok sounds good, I would suggest though to make that PR the next since it would be great to have it for the upcoming 0.3.0 release, for homogeneity with the changes that has been made in the core API.